frontend/Rhyolite/Frontend/Form.hs

```diff
@@ -83,49 +111,68 @@ tagPromptlyDynValidation (DynValidation (Compose b)) = attachPromptlyDynWithMayb
 manageValidity
   :: (DomBuilder t m, MonadHold t m, Prerender js m, PerformEvent t m)
   => Event t () -- When to validate
-  -> (Text -> Either Text a) -- Validation
+  -> (Behavior t Text -> BehaviorValidation t e a) -- Validation
+  -> (e -> Text) -- convert error to form for basic html validation
   -> m (InputElement EventResult (DomBuilderSpace m) t) -- Render input
-  -> m (InputElement EventResult (DomBuilderSpace m) t, DynValidation t Text a)
-manageValidity validate' validator renderInput = do
-  v@(input, val) <- manageValidation validate' validator renderInput
+  -> m (InputElement EventResult (DomBuilderSpace m) t, DynValidation t e a)
+manageValidity validate' validator errorText renderInput = do
+  v@(input, val) <- manageValidation validator renderInput
   prerender blank $ do
     let rawEl = _inputElement_raw input
     performEvent_ $ ffor (tagPromptlyDyn (fromDynValidation val) validate') $ \case
-      Left msg -> do
+      Left typedMsg -> do
+        let msg = errorText typedMsg
         setCustomValidity rawEl msg
         reportValidity_ rawEl
       _ -> setCustomValidity rawEl ("" :: Text) -- NOTE setting empty text is how the browser "clears" the error
   return v
 
+-- | The returned dynamic is really an abuse of I dynamic, namely it shouldn't
+-- be allowed to push anything. The one benefit of this vs 'fmap f . current' is
+-- the result can be fed into 'attachPromptly' rather than 'tag' to avoid
+-- loosing a frame.
+weirdMap
```

Comment:
This name... :joy: But surely there is a better name...

Reply:
good thing it's gone!